File: kotlin/day14_part1_2023.kt
import java.io.File

data class Coord(val x: Int, val y: Int) {
    fun add(c: Coord) = Coord(x + c.x, y + c.y)
    fun isInBounds(width: Int, height: Int) = x in 0 until width && y in 0 until height
}

data class Grid(val width: Int, val height: Int, val data: MutableMap<Coord, Char>)

val empty = '.'
val cubicRock = '#'
val roundRock = 'O'

val north = Coord(0, -1)
val west = Coord(-1, 0)
val south = Coord(0, 1)
val east = Coord(1, 0)

fun buildGrid(input: List<String>): Grid {
    val grid = Grid(input[0].length, input.size, mutableMapOf())

    input.forEachIndexed { y, line ->
        line.forEachIndexed { x, char ->
            if (char != empty) {
                grid.data[Coord(x, y)] = char
            }
        }
    }

    return grid
}

fun Grid.toString(): String {
    val result = StringBuilder()

    for (y in 0 until height) {
        for (x in 0 until width) {
            val coord = Coord(x, y)
            result.append(data[coord] ?: empty)
        }
        result.append("\n")
    }

    return result.toString()
}

fun shiftSingleRock(grid: Grid, coord: Coord, dir: Coord) {
    if (grid.data[coord] == roundRock) {
        var current = coord
        var before = coord.add(dir)

        while (!grid.data.containsKey(before) && before.isInBounds(grid.width, grid.height)) {
            grid.data[before] = roundRock
            grid.data.remove(current)

            current = before
            before = before.add(dir)
        }
    }
}

fun shiftRocks(grid: Grid, dir: Coord) {
    when (dir) {
        north, west -> {
            for (x in 0 until grid.width) {
                for (y in 0 until grid.height) {
                    shiftSingleRock(grid, Coord(x, y), dir)
                }
            }
        }
        south, east -> {
            for (x in grid.width - 1 downTo 0) {
                for (y in grid.height - 1 downTo 0) {
                    shiftSingleRock(grid, Coord(x, y), dir)
                }
            }
        }
    }
}

fun calculateLoad(grid: Grid): Int {
    var load = 0

    for (x in 0 until grid.width) {
        for (y in 0 until grid.height) {
            val coord = Coord(x, y)
            if (grid.data[coord] == roundRock) {
                load += grid.height - y
            }
        }
    }

    return load
}

fun solve(input: List<String>): Int {
    val grid = buildGrid(input)
    shiftRocks(grid, north)

    return calculateLoad(grid)
}

fun main() {
    val input = File("input.txt").readLines()
    println(solve(input))
}

File: go/day14_part1_2023.go
package main

import (
	"fmt"
	"os"
	"strings"
)

type Coord struct {
	X int
	Y int
}

func (c1 Coord) Add(c2 Coord) Coord {
	return Coord{c1.X + c2.X, c1.Y + c2.Y}
}

type Grid struct {
	Width  int
	Height int
	Data   map[Coord]byte
}

func (coord Coord) isInBounds(grid Grid) bool {
	return 0 <= coord.X && coord.X < grid.Width && 0 <= coord.Y && coord.Y < grid.Height
}

const (
	Empty     byte = '.'
	CubicRock byte = '#'
	RoundRock byte = 'O'
)

var (
	North = Coord{0, -1}
	West  = Coord{-1, 0}
	South = Coord{0, 1}
	East  = Coord{1, 0}
)

func buildGrid(input []string) Grid {
	grid := Grid{
		Width:  len(input[0]),
		Height: len(input),
		Data:   map[Coord]byte{},
	}

	for y, line := range input {
		for x, char := range line {
			if byte(char) != Empty {
				grid.Data[Coord{x, y}] = byte(char)
			}
		}
	}

	return grid
}

func (grid Grid) toString() string {
	var result string

	for y := 0; y < grid.Height; y++ {
		for x := 0; x < grid.Width; x++ {
			coord := Coord{X: x, Y: y}
			if v, ok := grid.Data[coord]; ok {
				result += string(v)
			} else {
				result += string(Empty)
			}
		}
		result += "\n"
	}

	return result
}

func shiftSingleRock(grid Grid, coord Coord, dir Coord) {
	if grid.Data[coord] == RoundRock {
		current := coord
		before := coord.Add(dir)

		_, ok := grid.Data[before]
		for !ok && before.isInBounds(grid) {
			grid.Data[before] = RoundRock
			delete(grid.Data, current)

			current = before
			before = before.Add(dir)
			_, ok = grid.Data[before]
		}
	}
}

func shiftRocks(grid Grid, dir Coord) {
	switch dir {
	case North, West:
		for x := 0; x < grid.Width; x++ {
			for y := 0; y < grid.Height; y++ {
				shiftSingleRock(grid, Coord{x, y}, dir)
			}
		}

	case South, East:
		for x := grid.Width - 1; x >= 0; x-- {
			for y := grid.Height - 1; y >= 0; y-- {
				shiftSingleRock(grid, Coord{x, y}, dir)
			}
		}
	}
}

func calculateLoad(grid Grid) int {
	load := 0

	for x := 0; x < grid.Width; x++ {
		for y := 0; y < grid.Height; y++ {
			coord := Coord{x, y}
			if grid.Data[coord] == RoundRock {
				load += grid.Height - y
			}
		}
	}

	return load
}

func solve(input []string) int {
	grid := buildGrid(input)
	shiftRocks(grid, North)

	return calculateLoad(grid)
}

func readFile(fileName string) []string {
	file, err := os.ReadFile(fileName)
	if err != nil {
		panic(err)
	}

	return strings.Split(strings.TrimSpace(string(file)), "\n")
}

func main() {
	input := readFile("input.txt")
	fmt.Println(solve(input))
}